### PR TITLE
Add dummy dataset management functionality

### PR DESCRIPTION
**Pull request by wattsecs**

### CODE REVIEW

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt]: This duplicates the functionality of putPicture. Couldn't we putPicture, then assign the resulting picture as representative on the category? KISS

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt]: does that cascade?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt]:
```suggestion
    suspend fun deletePicture(picture: CategorizedPicture)
```

**Review by bbjubjub2494 — COMMENTED** [review]
some feedback. Also we should draw an ER model.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt]: Thank you for noticing, I guess I forgot that one

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt]: Ultimately yes in my opinion. A Category is always the same (e.g pictures of forks are pictures of forks, no matter the dataset). This way categories can be crowd sourced. Support for creating Categories outside of datasets could be added later.  
Obviously support for categories personnal to users, maintainers for categories etc could (and should imo) be added later.
This decision is up to debate though.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt]: Since we have no clear definition of a dataset I put it like this for now, I think that adding one ~~and a corresponding matcher would be a better choice though.~~

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt]: To be coherent with https://github.com/HumanLearning2021/HumanLearningApp/pull/50#discussion_r598807381 my idea would be more of an "unlink" (keep the category and the pictures but don't use them in this specific dataset). The function could be renamed though to make this clearer.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt]: ~~For now this is the best way I found. See https://github.com/HumanLearning2021/HumanLearningApp/pull/50#discussion_r598808597~~
True, this will be changed.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/DatasetsManagement.kt]: It will be a singleton, sorry.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt]: ~~I envisioned a pop-up message which makes the user decide whether he wants to delete it or move it into the other pictures since a representative pictures.~~
I want to give the option to either remove it or move it to the other pictures. This will need an overload.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt]: Yes, an overload with `putRepresentativePicture(picture: CategorizedPicture)`. This way the representative can be chosen from an already existing picture or be taken from the camera.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt]: See https://github.com/HumanLearning2021/HumanLearningApp/pull/50#discussion_r598807381

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DatasetsManagementPresenterInterface.kt
Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DatasetsManagementPresenterInterface.kt]: All local datasets? All datasets in the database ? Might need some precisions

**Review by Nitwix — COMMENTED** [review]
Some minor comments/questions, but otherwise, the interfaces seem well defined and useful

File: .idea/compiler.xml
Comment (wattsecs) [reply .idea/compiler.xml]: For some reason reason the project would not build otherwise (neither locally, nor on cirrus)

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DatasetsManagementPresenterInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DatasetsManagementPresenterInterface.kt]: All the datasets in the database. For now the datasets, categories and pictures are considered as shared between all users. Restrictions can (and will) be added in later sprints.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt
Comment (NicolasVial) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt]: This will give us all the pictures of the same category in this dataset right? If so, it's perfect for me.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt
Comment (NicolasVial) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt]: When we delete a Category in a dataset, it will also "delete" (more like remove from this dataset than really remove tho) all the pictures that belong to this category in the dataset?

**Review by NicolasVial — COMMENTED** [review]
It looks perfect for our actual plan. Some stuff will have to be clarified but this is going to be in later sprints. Nice work.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt]: Yes exactly

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt]: It is more of an "unlink". The category will be kept. See https://github.com/HumanLearning2021/HumanLearningApp/pull/50#discussion_r598807381

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagement.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagement.kt]: This might not be an object. Worst case scenario I'll change it though since this fact is only relevant at dependency injection time.

**Review by Jaakik — APPROVED** [review]
Looks good to me 👍🏽

**Review by martin-lenweiter — COMMENTED** [review]
Looks good! Let's have another review when what was discussed in the chat is implemented

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt]: Good idea.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt]: See https://github.com/HumanLearning2021/HumanLearningApp/pull/50#discussion_r598807381

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt]: 1. clients should not generate unique keys themselves
2. What happened to human-readable name?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt]:
```suggestion
    fun deleteDataset(dataset: Dataset)
```

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt]: True, mistake on my part. I will put it back as human-readable dataset name

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DatasetsManagementInterface.kt]: We don't have a Dataset class?

**Comment by bbjubjub2494** [comment]
Is the current state roughly what you want in master or do you intend to implement what was discussed on discord re database and ER model?

**Comment by wattsecs** [comment]
> 
> 
> Is the current state roughly what you want in master or do you intend to implement what was discussed on discord re database and ER model?

These are the methods that are needed and that I want to have in master. In my opinion, changing the way we access the database should be changed as discussed on Discord, yes.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt]: [MutableSet is a subclass of Set](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-mutable-set/#kotlin.collections.MutableSet), you don't need to change it in the interface. please revert this.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt]: Isn't the whole point of `kotlin-parcelize` to not have all of this boilerplate?

**Comment by bbjubjub2494** [comment]
please nuke `interface DatasetInterface`

**Comment by wattsecs** [comment]
> 
> 
> please nuke `interface DatasetInterface`

Thanks, I missed that one.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt]: I am not entirely sure. I will look into it.

Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt]: It does indeed, thanks!

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt
Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt]: useful comment, but misplaced, no?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt
Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt:15]: Cool idea!

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt
Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt:23]: Is this necessary?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt
Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt:46]: maybe consider factoring this out (as suggested by codeclimate)

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseService.kt
Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseService.kt]:
```suggestion
        throw IllegalArgumentException("The dataset with id $id is not present in the database")
```

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt
Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt:25]: isn't this redundant?

Comment (Nitwix) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt:25]: Isn't this redundant?

**Review by Nitwix — APPROVED** [review]
Truly excellent work! It will allow us to integrate a lot of things.
I left some questions and comments

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/DisplayDatasetActivityTest.kt
Comment (wattsecs) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/DisplayDatasetActivityTest.kt:87]: Thanks, I totally forgot about that 😅

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyCategorizedPictureTest.kt
Comment (wattsecs) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyCategorizedPictureTest.kt:53]: I tried that but I did not work. All the ways I found on the Internet did not either...

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatabaseService.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatabaseService.kt:90]: Yes because it allows to be more flexible and since the ids are never provided by the "users"  ids used will always be of the type expected by the implementation.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatabaseService.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatabaseService.kt:124]: Yes but it will be done at the same time as all the other Firestore implementations.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt]: I would argue no because the design of the other interfaces depends on it

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt:23]: Yes because I decided not to deal with the exceptions in the dummy implementation

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt:15]: Thanks

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt:46]: I think it is too complex to make it into a useful method to warrant the effort

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseService.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseService.kt]: Thanks

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt
Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt:25]: See https://github.com/HumanLearning2021/HumanLearningApp/pull/50#discussion_r606379453

Comment (wattsecs) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt:25]: See https://github.com/HumanLearning2021/HumanLearningApp/pull/50#discussion_r606379453

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt
Comment (Nitwix) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt]: No I mean it should be just above the attribute (look at the file for context)

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt
Comment (Nitwix) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDatabaseManagement.kt:23]: but couldn't you remove the whole try/catch block? It would be equivalent no?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt
Comment (Nitwix) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/Dataset.kt]: Oh ok, I see what you mean.

**Pull request merged**
